website/docs/API.md

```diff
@@ -515,22 +517,96 @@ Defined as:
 ```jsx
 function waitFor<T>(
   expectation: () => T,
-  { timeout: number = 4500, interval: number = 50 }
+  { timeout: number = 1000, interval: number = 50 }
 ): Promise<T> {}
 ```
 
-Waits for non-deterministic periods of time until your element appears or times out. `waitFor` periodically calls `expectation` every `interval` milliseconds to determine whether the element appeared or not.
+Should be used when testing asynchronous code or code with timers. `waitFor` allows to wait for some expectation (e.g. the presence of an element on the screen). The `expectation` param is a callback that should throw until what you're waiting for has realized.
+
+For instance, this is how we'd wait for an element to appear:
+
+```tsx
+await waitFor(() => expect(screen.getByText('Banana ready').toBeTruthy()))
+```
+
+What `waitFor` does is executing `expectation()` at intervals of duration equal to the optional `interval` param (default is 50ms) until it doesn't throw anymore. It will throw if it reaches its timeout, defined by the optional `timeout` param (default is 1000ms). If `expectation()` does not throw then `waitFor` will resolve right away.
+
+```tsx
+// won't throw
+await waitFor(() => false);
+```
+
+`waitFor` only rejects when it reaches its timeout so if you don't await it it will never throw.
 
 ```jsx
-import { render, screen, waitFor } from '@testing-library/react-native';
+// won't throw
+waitFor(() => expect(1).toBe(2))
+```
 
-test('waiting for an Banana to be ready', async () => {
-  render(<Banana />);
+:::note
+You can enforce awaiting waitFor by using the [await-async-utils](https://github.com/testing-library/eslint-plugin-testing-library/blob/main/docs/rules/await-async-utils.md) rule from [eslint-plugin-testing-library](https://github.com/testing-library/eslint-plugin-testing-library).
+:::
 
-  await waitFor(() => screen.getByText('Banana ready'));
-});
+Since `waitFor` is likely to run `expectation()` multiple times, it is highly recommended not to have any side effects in `waitFor`.
+
+```jsx
+await waitFor(() => {
+  // button will be pressed on each waitFor iteration
+  fireEvent.press(screen.getByText('press me'))
+  expect(screen.getByText('pressed')).toBeTruthy()
+})
 ```
 
+:::note
+Not using fireEvent in `waitFor` be enforced with the rule [no-wait-for-side-effects](https://github.com/testing-library/eslint-plugin-testing-library/blob/main/docs/rules/no-wait-for-side-effects.md).
+:::
+
+### waitFor with jest fake timers
+
+`waitFor` checks wether jest fake timers are enabled and adapts its behavior when it is the case. It will create a variable counting the time remaining which initially equals `timeout` and is decreased on each iteration by `interval`. When that variable is negative it will reject. Additionnaly, on each iteration it will call `jest.advanceTimersByTime(interval)`.
```

Comment:
I think your simplified code below is enough to understand what's happening, no need to paraphrase it here

Reply:
Agreed, I reworked this part a bit, I think it's better, tell me how you feel about it

Reply:
much better!